bedrock/3_EntityWorkflow/workflows.py

```diff
@@ -0,0 +1,174 @@
+    def __init__(self) -> None:
+        # List to store prompt history
+        self.conversation_history: List[Tuple[str, str]] = (
+            []
+        )
+        self.prompt_queue: Deque[str] = deque()
+        self.conversation_summary: Optional[str] = None
+        self.continue_as_new_per_turns: int = 6
+        self.chat_ended: bool = False
+
+    @workflow.run
+    async def run(
+        self,
+        params: BedrockParams = None,
+    ) -> str:
+
+        if params and params.conversation_summary:
+            self.conversation_history.append(
+                ("conversation_summary", params.conversation_summary)
+            )
+
+        if params and params.prompt_queue:
+            for prompt in params.prompt_queue:
+                self.prompt_queue.append(prompt)
+
+        summary_activity_task: Optional[asyncio.Task] = None
+
+        while True:
+            workflow.logger.info("Waiting for prompts...")
+
+            # Wait for a chat message (signal) or timeout
+            await workflow.wait_condition(
+                lambda: bool(self.prompt_queue) or self.chat_ended
+            )
+
+            # if end chat signal was sent
+            if self.chat_ended:
+                # the workflow might be continued as new without any
+                # chat to summarize
+                if summary_activity_task is not None:
+                    # ensure conversation summary task has finished
+                    # before closing the workflow (avoid race)
+                    await workflow.wait_condition(
+                        lambda: summary_activity_task.done()
+                    )
+                else:
+                    # conversation history from previous workflow
+                    self.conversation_summary = params.conversation_summary
+
+                workflow.logger.info(
+                    "Chat ended. Conversation summary:\n" +
+                    f"{self.conversation_summary}"
+                )
+
+                return f"{self.conversation_history}"
+
+            # Fetch next user prompt and add to conversation history
+            prompt = self.prompt_queue.popleft()
+            self.conversation_history.append(("user", prompt))
+
+            workflow.logger.info("Prompt: " + prompt)
+
+            # Send prompt to Amazon Bedrock
+            response = await workflow.execute_activity(
+                prompt_bedrock,
+                self.prompt_with_history(prompt),
+                schedule_to_close_timeout=timedelta(seconds=20),
+            )
+
+            workflow.logger.info(f"{response}")
+
+            # Append the response to the conversation history
+            self.conversation_history.append(("response", response))
+
+            # summarize the conversation to date using Amazon Bedrock
+            # uses start_activity with a callback
+            # so it doesn't block new messages being sent to Amazon Bedrock
+            summary_activity_task = workflow.start_activity(
```

Comment:
So you are summarizing after every user prompt? If I send 20 user prompts, that's 20 summary calls? If this is a requirement, arguably no value in having it be separate activities then (the same activity can issue single prompt and request summary w/ full history).

Reply:
Have fixed this to be consistent with the signals and queries sample - where summarize is only called at conversation end. Although in this case it's also summarized on CAN.